pkg/controller/pkg/revision/deployment.go

```diff
@@ -133,5 +146,20 @@ func buildProviderDeployment(provider *metav1alpha1.Provider, revision v1alpha1.
 			d.Spec.Template.Spec.Containers[0].Env = cc.Spec.Env
 		}
 	}
-	return s, d
+	if err := ensureLeaderElection(d); err != nil {
+		return nil, nil, err // replicas > 1 but LEADER_ELECTION is not set or is not 'true'
+	}
```

Comment:
@negz @hasheddan I've decided to error out if the provided replicas is more than `1` and leader election is not enabled. This prevents a user to make a mistake and leaves a log trace of what happened for them to view and fix. What do you think?

Reply:
Hmm I'm not sure we should enforce this requirement. There are a variety of ways a user could misconfigure a controller deployment here and it feels more future proof to not lock in provider authors to a specific environment variable.

Reply:
I'm not sure if I got your point completely. This restriction is only for number of replica (which is 1 by default). It means if a user intentionally wants to bump it for a provider they need to use the new `ControllerConfig` you added, and by convention they need to also pass the extra environment variable needed for this.

```yaml
apiVersion: pkg.crossplane.io/v1alpha1
kind: ControllerConfig
metadata:
  name: aws-config
spec:
  replicas: 2
  env:
    - name: "LEADER_ELECTION"
      value: "true"
```

I originally had the idea to auto-populate the environment (if missing) here if `replicas > 1`, but thought this might feel more like a automagic to the user rather than throwing an actual error for it. This can be described (e.g. in a README or https://doc.crds.dev/ maybe).

Reply:
Considering the CRD reads:

```go
type ControllerConfigSpec struct {
	// Number of desired pods. This is a pointer to distinguish between explicit
	// zero and not specified. Defaults to 1.
	// Note: If number more than 1 is set, environment variable "LEADER_ELECTION"
	// must exist and must be true.
	// +optional
	Replicas *int32 `json:"replicas,omitempty"`
...
}
```

Reply:
> // Note: If number more than 1 is set, environment variable "LEADER_ELECTION"
	// must exist and must be true.

My point is that I don't think we should tie replicas to a specific environment variable being set. If a provider wanted to use a different environment variable, or they had leader election enabled by default, then a user may just want to set a replica number without the environment variable. It is true that a user could accidentally supply a bad configuration that caused controllers to fight with each other, but I don't think we should explicitly restrict that from happening. Ensuring this environment variable is set doesn't even guarantee that the configuration is valid (i.e. a provider may not honor `LEADER_ELECTION`), and in that case, enforcing this could actually cause additional confusion.

Reply:
Oh I see your point! That makes sense. By this approach I was trying to implement a way to pass input corresponding to [this](https://github.com/crossplane/provider-aws/pull/408/files#diff-36b6d20eb5aea66cf39f8d94111bd96513626ef7f61459f0d9e8e9507ded1d17R39) line. I've looked around a bit and the only options I could have thought were:

- pass it as environment variable to container
- pass it as `args` to container

Both cases needed a change of provider's `Deployment` and environment variable was a safer approach.

Now, I'm thinking maybe we should remove the flag from providers and move the leader election enablement to the providers (i.e. enabled by default or disabled altogether).

Reply:
I agree with @hasheddan here; we should allow `ControllerConfig` authors to specify their desired number of replicas regardless of whether or not they have enabled leader election (either by setting the `LEADER_ELECTION` env var or some other means). I think this is a place where we should rely on convention (i.e. expect users to provide correct configuration) rather than enforcement.

So, I like the pattern you've established in https://github.com/crossplane/provider-aws/pull/408 (and similar PRs) that adds a flag and env var that can be used to enable leader election; I think we should keep that. I think the only change we should make is to remove the `ensureLeaderElection` check here.

Reply:
Great, will update. On the same subject, to be more feature complete and flexible, maybe we can add `Args` to `Spec.Template.Spec.Containers` too? This will not be a blocker of this PR but since the file is being change I can add it too and rename the title of the PR too.

Reply:
@khos2ow I would support adding `Args` :) Would you like to do it in this PR or in a follow-up?

Reply:
In this particular case, with only one line change needed in the same file of this PR I prefer to do it here, otherwise opening a follow-up PR is completely fine by me if you prefer that.

Reply:
Here is great :)